### ISSUE TITLE
expert_bias is updated during training but saved checkpoint contains only zero values

### ISSUE THREAD

**Issue by trestad** [pytorch/torchtitan#1222]
### Bug description

Hi Torchtitan team,

I've encountered an issue while training LLaMA 4. During training, I observed that the expert_bias parameters are being updated as expected — I can confirm this by printing their values in the MoE's forward function, and they do change over time.

However, when I save a checkpoint, the stored expert_bias tensors are all zeros. When I inspect the checkpoint manager's model states dict just before saving, I can see that expert_bias is already zero at that point (torchtitan/components/checkpoint.py L380), which implies the issue occurs before the actual save.

My setup follows experiment/llama4/train_configs, except for the model architecture and training hyperparameters.

I haven’t yet traced the exact origin of the issue but I’m wondering has the team verified the saved parameter contents for LLaMA 4? Or, has anyone else in the community encountered a similar issue?

Any insights or suggestions for where this might be going wrong would be much appreciated!

Thanks!

### Versions

torch 2.8.0, latest torchtitan (May 22nd), Below is my toml config file:

[job]
dump_folder = "outputs/checkpoints/1b-3of32"
description = "1bx64e"
print_args = true
use_for_integration_test = true

[profiling]
enable_profiling = false
save_traces_folder = "profile_trace"
profile_freq = 10
enable_memory_snapshot = false
save_memory_snapshot_folder = "memory_snapshot"

[comm]
init_timeout_seconds = 3600
train_timeout_seconds = 3600

[metrics]
log_freq = 1
disable_color_printing = false
enable_tensorboard = false
save_tb_folder = "tb"
enable_wandb = false

[model]
name = "llama4"
flavor = "1b-3of32"
tokenizer_path = "llama3_tokenizer.model"

[optimizer]
name = "AdamW"
lr = 1e-3
eps = 1e-15

[lr_scheduler]
warmup_steps = 1000
decay_type = "cosine"
lr_min = 0.1

[training]
batch_size = 2
accumulate_steps = 1
seq_len = 2048
max_norm = 1.0  # grad norm clipping
steps = 40000
compile = false
dataset_type = "hf"  # mmap for megatron style
dataset = "huggingfacefw/fineweb-edu"
dataset_path = "/pretrain_data/fineweb-edu"
loss_function_type = "cross_entropy"
seed = 1234

[parallelism]
data_parallel_replicate_degree = 1
data_parallel_shard_degree = -1
fsdp_reshard_after_forward = "default" # default / never / always
tensor_parallel_degree = 1
enable_async_tensor_parallel = false
pipeline_parallel_degree = 1
context_parallel_degree = 1

[checkpoint]
enable_checkpoint = true
folder = "checkpoint"
interval = 1
model_weights_only = false
keep_latest_k = 0
export_dtype = "float32"
async_mode = "disabled"  # ["disabled", "async", "async_with_pinned_mem"]

[activation_checkpoint]
mode = "none"  # ["none", "selective", "full"]

[float8]
enable_fsdp_float8_all_gather = false
precompute_float8_dynamic_scale_for_fsdp = false
filter_fqns = ["output", "router.gate"]

**Comment by trestad** [comment]
Hi Torchtitan team,

Thanks again for your work on this project.

I wanted to update this issue — I’ve identified and resolved the root cause.

The problem was that expert_bias was originally registered as a buffer via register_buffer, but later in the _update_expert_bias() function, it was being reassigned like this:


`self.expert_bias = self.expert_bias + expert_bias_delta`

After reassignment, the original expert_bias is no longer included in checkpointer.states['model'].state_dict() or the saved checkpoint, which explains why the saved values were always zero.

The fix was simply to update the buffer in-place, like this:

`self.expert_bias.add_(expert_bias_delta)`

The buffer is now correctly updated and saved in the checkpoints. It would be great to get this fixed, as it can be a subtle but impactful bug. Hopefully this comment will also help others who might run into the same issue.

Thanks again!
